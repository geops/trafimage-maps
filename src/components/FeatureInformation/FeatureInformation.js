import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import Button from '@geops/react-ui/components/Button';
import { MdClose } from 'react-icons/md';
import { IoIosArrowRoundBack, IoIosArrowRoundForward } from 'react-icons/io';
import { setClickedFeatureInfo } from '../../model/app/actions';
import popups from '../../popups';

import './FeatureInformation.scss';

const propTypes = {
  clickedFeatureInfo: PropTypes.array.isRequired,
};

const FeatureInformation = ({ clickedFeatureInfo }) => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [featureIndex, setFeatureIndex] = useState(0);

  useEffect(() => {
    setFeatureIndex(0);
  }, [clickedFeatureInfo]);

  const features = clickedFeatureInfo.map(l => l.features).flat();
  const feature = features[featureIndex];
  if (!feature) {
    return null;
  }

  const info = clickedFeatureInfo.find(i => i.features.includes(feature));
  if (!info || !info.layer) {
    return null;
  }

  const PopupComponent = popups[info.layer.get('popupComponent')];
  if (!PopupComponent) {
    return null;
  }

  let pagination = null;
  const { layer } = info;

  if (features.length > 1) {
    pagination = (
      <div className="wkp-pagination-wrapper">
        <span className="wkp-pagination-button-wrapper">
          {featureIndex > 0 ? (
            <Button
              className="wkp-pagination-button"
              onClick={() => setFeatureIndex(featureIndex - 1)}
            >
              <IoIosArrowRoundBack />
            </Button>
          ) : null}
        </span>
        {featureIndex + 1} {t('von')} {features.length}
        <span className="wkp-pagination-button-wrapper">
          {featureIndex + 1 < features.length ? (
            <Button
              className="wkp-pagination-button"
              onClick={() => setFeatureIndex(featureIndex + 1)}
            >
              <IoIosArrowRoundForward />
            </Button>
          ) : null}
        </span>
      </div>
    );
  }
  return (
    <div className="wkp-feature-information">
      <React.Suspense fallback="...loading">
        <div className="wkp-feature-information-header">
          <span>
            {PopupComponent && PopupComponent.renderTitle && feature
              ? PopupComponent.renderTitle(feature)
              : layer && layer.getName() && t(layer.getName())}
          </span>
          <Button
            className="wkp-close-bt"
            onClick={() => {
              dispatch(setClickedFeatureInfo());
              if (PopupComponent.onCloseBtClick) {
                PopupComponent.onCloseBtClick();
              }
            }}
          >
            <MdClose focusable={false} />
          </Button>
        </div>
        <div className="wkp-feature-information-body">
          <PopupComponent
            key={info.layer.getKey()}
            feature={features[featureIndex]}
          />
          {pagination}
        </div>
      </React.Suspense>
    </div>
  );
};

FeatureInformation.propTypes = propTypes;

export default React.memo(FeatureInformation);
