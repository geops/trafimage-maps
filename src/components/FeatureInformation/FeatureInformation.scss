.wkp-feature-information {
  .wkp-feature-information-header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    background-color: #f5f5f5;
    font-weight: bold;
    height: 40px;

    span {
      display: flex;
      align-items: center;
      padding-left: 10px;
      height: 100%;
      overflow: hidden;
    }

    .wkp-close-bt {
      display: flex;
      align-items: center;
      justify-content: center;
      flex: 0 0 40px;
      height: 100%;
    }
  }

  .wkp-feature-information-body {
    padding: 10px;

    .wkp-pagination-wrapper {
      display: flex;
      align-items: center;
      justify-content: center;
      padding-top: 5px;

      .wkp-pagination-button-wrapper {
        padding: 0 10px;
        height: 25px;
        width: 25px;

        .wkp-pagination-button {
          height: 25px;
          width: 25px;

          svg {
            height: 25px;
            width: 25px;
          }
        }
      }
    }
  }
}
