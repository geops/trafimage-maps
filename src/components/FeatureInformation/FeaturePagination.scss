.wkp-pagination-wrapper {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 35px;

  .wkp-pagination-button-wrapper {
    padding: 0 10px;
    height: 25px;
    width: 25px;

    .wkp-pagination-button {
      height: 25px;
      width: 25px;
      cursor: pointer;

      svg {
        height: 25px;
        width: 25px;
      }
    }
  }
}
