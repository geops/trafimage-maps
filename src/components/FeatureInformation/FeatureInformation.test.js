import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Feature } from 'ol';
import { Point } from 'ol/geom';
import Layer from 'react-spatial/layers/Layer';

import FeatureInformation from '.';

describe('FeatureInformaion', () => {
  const mockStore = configureStore([thunk]);
  let store;

  beforeEach(() => {
    store = mockStore({
      map: {},
      app: { projection: { value: 'EPSG:3857' } },
      oidc: { user: {} },
    });
  });

  describe('should match snapshot.', () => {
    test("using the layers's popupComponent", () => {
      const l = new Layer({
        key: 'foo',
        properties: {
          popupComponent: 'NetzkartePopup',
        },
      });
      const fi = [
        {
          features: [new Feature(new Point([2, 2]))],
          layer: l,
        },
      ];

      const component = renderer.create(
        <Provider store={store}>
          <FeatureInformation featureInfo={fi} />
        </Provider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test("using the info's popupComponent", () => {
      const l = new Layer({
        key: 'foo',
      });
      const fi = [
        {
          features: [new Feature(new Point([2, 2]))],
          layer: l,
          popupComponent: 'NetzkartePopup',
        },
      ];

      const component = renderer.create(
        <Provider store={store}>
          <FeatureInformation featureInfo={fi} />
        </Provider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should display multiple features with pagination', () => {
      const l = new Layer({
        key: 'foo',
      });
      const fi = [
        {
          features: [
            new Feature(new Point([2, 2])),
            new Feature(new Point([1, 1])),
          ],
          layer: l,
          popupComponent: 'NetzkartePopup',
        },
        {
          features: [
            new Feature(new Point([2, 2])),
            new Feature(new Point([1, 1])),
          ],
          layer: l,
          popupComponent: 'NetzkartePopup',
        },
      ];

      const component = renderer.create(
        <Provider store={store}>
          <FeatureInformation featureInfo={fi} />
        </Provider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should not display header', () => {
      const l = new Layer({
        key: 'foo',
      });
      const fi = [
        {
          features: [new Feature(new Point([2, 2]))],
          layer: l,
          popupComponent: 'KilometragePopup',
        },
      ];

      const component = renderer.create(
        <Provider store={store}>
          <FeatureInformation featureInfo={fi} />
        </Provider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
