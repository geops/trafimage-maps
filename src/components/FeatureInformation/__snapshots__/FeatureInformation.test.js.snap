// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureInformaion should match snapshot. should display multiple features with pagination 1`] = `
<div
  className="wkp-feature-information"
>
  <div
    className="wkp-feature-information-header"
  >
    <span />
    <div
      aria-label="Popup schliessen"
      className="wkp-close-bt"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
      title="Popup schliessen"
    >
      <svg
        alt="Popup schliessen"
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
  </div>
  <div
    className="wkp-feature-information-body"
  >
    <div
      className="wkp-netzkarte-popup"
    >
      <div>
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Abfahrtszeiten
        </div>
      </div>
      <div>
        <div
          aria-expanded={false}
          className="wkp-coordinate-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Koordinaten
        </div>
        <div
          className="wkp-coordinates-infos wkp-closed"
        >
          <span
            className="wkp-projection-label"
          />
          <span>
            Länge (X): 2
          </span>
          <span>
            Breite (Y): 2
          </span>
        </div>
      </div>
    </div>
    <div
      className="wkp-pagination-wrapper"
    >
      <span
        className="wkp-pagination-button-wrapper"
      />
      1
       
      von
       
      4
      <span
        className="wkp-pagination-button-wrapper"
      >
        <div
          aria-label="weiter"
          className="wkp-pagination-button"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
          title="weiter"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M295.6 163.7c-5.1 5-5.1 13.3-.1 18.4l60.8 60.9H124.9c-7.1 0-12.9 5.8-12.9 13s5.8 13 12.9 13h231.3l-60.8 60.9c-5 5.1-4.9 13.3.1 18.4 5.1 5 13.2 5 18.3-.1l82.4-83c1.1-1.2 2-2.5 2.7-4.1.7-1.6 1-3.3 1-5 0-3.4-1.3-6.6-3.7-9.1l-82.4-83c-4.9-5.2-13.1-5.3-18.2-.3z"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`FeatureInformaion should match snapshot. should not display header 1`] = `
<div
  className="wkp-feature-information"
>
  <div
    className="wkp-feature-information-body"
  >
    <div
      className="wkp-kilometrage-popup"
    >
      <div
        className="wkp-popup-row"
      >
        DfA Linien Nr.: undefined
      </div>
      <div
        className="wkp-popup-row"
      >
        Kilometer: NaN
      </div>
    </div>
  </div>
</div>
`;

exports[`FeatureInformaion should match snapshot. using the info's popupComponent 1`] = `
<div
  className="wkp-feature-information"
>
  <div
    className="wkp-feature-information-header"
  >
    <span />
    <div
      aria-label="Popup schliessen"
      className="wkp-close-bt"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
      title="Popup schliessen"
    >
      <svg
        alt="Popup schliessen"
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
  </div>
  <div
    className="wkp-feature-information-body"
  >
    <div
      className="wkp-netzkarte-popup"
    >
      <div>
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Abfahrtszeiten
        </div>
      </div>
      <div>
        <div
          aria-expanded={false}
          className="wkp-coordinate-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Koordinaten
        </div>
        <div
          className="wkp-coordinates-infos wkp-closed"
        >
          <span
            className="wkp-projection-label"
          />
          <span>
            Länge (X): 2
          </span>
          <span>
            Breite (Y): 2
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FeatureInformaion should match snapshot. using the layers's popupComponent 1`] = `
<div
  className="wkp-feature-information"
>
  <div
    className="wkp-feature-information-header"
  >
    <span />
    <div
      aria-label="Popup schliessen"
      className="wkp-close-bt"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
      title="Popup schliessen"
    >
      <svg
        alt="Popup schliessen"
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
  </div>
  <div
    className="wkp-feature-information-body"
  >
    <div
      className="wkp-netzkarte-popup"
    >
      <div>
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Abfahrtszeiten
        </div>
      </div>
      <div>
        <div
          aria-expanded={false}
          className="wkp-coordinate-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Koordinaten
        </div>
        <div
          className="wkp-coordinates-infos wkp-closed"
        >
          <span
            className="wkp-projection-label"
          />
          <span>
            Länge (X): 2
          </span>
          <span>
            Breite (Y): 2
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
